File: rust/src/leetcode/Q518_coin_change_ii.rs
/*
 * @lc app=leetcode id=518 lang=rust
 *
 * [518] Coin Change II
 */

// @lc code=start
impl Solution {
    pub fn change(amount: i32, coins: Vec<i32>) -> i32 {
        let mut dp: Vec<Vec<i32>> = vec![vec![0; amount as usize + 1]; coins.len() + 1];

        for i in 0..=coins.len() {
            dp[i][0] = 1;
        }

        for i in 1..=coins.len() {
            for j in 1..=amount as usize {
                let coin: i32 = coins[i - 1];

                if j < coin as usize {
                    dp[i][j] = dp[i - 1][j];
                } else {
                    dp[i][j] = dp[i - 1][j] + dp[i][j - coin as usize];
                }
            }
        }

        dp[coins.len()][amount as usize]
    }
}
// @lc code=end

File: python/Q518_coin_change_ii.py
#
# @lc app=leetcode id=518 lang=python3
#
# [518] Coin Change II
#


# @lc code=start
class Solution:
    def change(self, amount: int, coins: list[int]) -> int:
        dp: list[list[int]] = [[0] * (amount + 1) for _ in range(len(coins) + 1)]
        for i in range(len(coins) + 1):
            dp[i][0] = 1

        for i in range(1, len(coins) + 1):
            for j in range(1, amount + 1):
                coin: int = coins[i - 1]

                if j < coin:
                    dp[i][j] = dp[i - 1][j]
                else:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - coin]

        return dp[-1][-1]


# @lc code=end
